### ISSUE TITLE
ANRs when getString() is called

### ISSUE THREAD

**Issue by JulienDev** [phrase/phrase-android#37]
Hello,

The Google Play Store console is reporting us a lot ANRs (around one million each month) that seems to be due to Android Phrase SDK. I've already reported this to your technical support but never got any answer.
Each ANR happen when a broadcast is received or when a service is started by the app, so it's possible the issue is only there in background. The stacktrace of each report is always the same and look like that :

```
  at sun.misc.Unsafe.park (Native method)
  at java.util.concurrent.locks.LockSupport.park (LockSupport.java:190)
  at java.util.concurrent.FutureTask.awaitDone (FutureTask.java:450)
  at java.util.concurrent.FutureTask.get (FutureTask.java:192)
  at com.phrase.android.sdk.RepositoryHolder.getRepository (RepositoryHolder.java:1)
  at com.phrase.android.sdk.inject.PhraseResources.b (SourceFile:3)
  at com.phrase.android.sdk.inject.PhraseResources.getText (SourceFile:1)
  at com.phrase.android.sdk.inject.PhraseResources.getString (SourceFile:1)
  at android.content.Context.getString (Context.java:649)
```

ANRs have an impact on our app quality and Play Store app ranking, that's why we want to get rid of them ASAP. Could you please look at it and let me know how to fix it?

Thanks

**Comment by theSoenke** [comment]
Hey @JulienDev, we're looking into this right now and I'll try to provide an update as soon as possible

**Comment by angelolloqui** [comment]
Hi, we have the same issue in production. When we make use of the first `getString` it creates a ANR in some cases. I think it is waiting for the download to complete, but I wonder why not just returning the binary string while the request is ongoing instead of waiting? @theSoenke Do you have an estimated time for resolving this issue?

**Comment by timoEmberNjiuko** [comment]
@JulienDev Hi, this seems a bit late, but do you still have these ARNs or a more detailed Stack trace avaiblable? This would be huge help to find the cause :) Thanks in advance!

**Comment by timoEmberNjiuko** [comment]
Hi @angelolloqui. Could you provide some more information here? :) Does it occurs only on the first getString? Do you got an full stack trace for us? :) Whats your case, when this occurs? Thanks in advance!

**Comment by angelolloqui** [comment]
Here it is some random ANR stack trace:

```
main (waiting): tid=1 systid=5725 
       at sun.misc.Unsafe.park(Native method)
       at java.util.concurrent.locks.LockSupport.park(LockSupport.java:190)
       at java.util.concurrent.FutureTask.awaitDone(FutureTask.java:450)
       at java.util.concurrent.FutureTask.get(FutureTask.java:192)
       at com.phrase.android.sdk.RepositoryHolder.getRepository(SourceFile:1)
       at com.phrase.android.sdk.inject.PhraseResources.b(SourceFile:3)
       at com.phrase.android.sdk.inject.PhraseResources.getText(SourceFile:1)
       at com.phrase.android.sdk.inject.PhraseResources.getString(SourceFile:1)
       at android.content.Context.getString(Context.java:705)
       at com.playtomic.general.DependencyProvider.<init>(DependencyProvider.kt:82)
       at com.playtomic.general.DependencyProvider.<init>(DependencyProvider.kt:58)
       at com.playtomic.general.MainActivity.createDependencies(MainActivity.kt:108)
       at com.playtomic.general.MainActivity.onCreate(MainActivity.kt:47)
       at android.app.Activity.performCreate(Activity.java:8129)
       at android.app.Activity.performCreate(Activity.java:8109)
       at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1344)
       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3750)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3943)
       at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
       at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
       at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2348)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loopOnce(Looper.java:233)
       at android.os.Looper.loop(Looper.java:344)
       at android.app.ActivityThread.main(ActivityThread.java:8191)
       at java.lang.reflect.Method.invoke(Native method)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:584)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1034)
```

In line `DependencyProvider.kt:82` we have this:
```
activity.getString(R.string.app_name)
```
What I see from the code is that this usage is happening before the call to `Phrase.updateTranslations()` is made. So in my app, the flow is:
1. `Application.onCreate()` -> `Phrase.setup`
2. `MainActivity.onCreate()` -> `activity.getString(R.string.app_name)`
3. `MainActivity.onCreate()` -> `Phrase.updateTranslations()`

<img width="452" alt="Screenshot 2022-06-09 at 10 02 07" src="https://user-images.githubusercontent.com/1382659/172796510-a200eca1-8634-4d71-a098-0ce9866c41e3.png">

I see no patter on devices (a lot of Samsung but not that rare due to our user base), only Android 11 and 12 apparently, and not sure how to interpret the background state since half are reported in background but this  is happening when starting the app.

Not sure if this helps.

**Comment by timoEmberNjiuko** [comment]
@angelolloqui Huge thanks for the help. I will have a look into it 👍

**Comment by RogerParis** [comment]
@angelolloqui does it make any difference if you put the `Phrase.updateTranslations()` code in your `Application.onCreate()`? as stated in the documentation here: https://help.phrase.com/help/android-sdk-installation-instructions

**Comment by RogerParis** [comment]
Also you could try, if you didn't, to add the documentation step: 

`public class BaseActivity extends AppCompatActivity {
  @NonNull
  @Override
  public AppCompatDelegate getDelegate() {
    return Phrase.getDelegate(this, super.getDelegate());
  }
}`

So then in the activity you can call `applicationContext.getString(R.string.app_name)` instead of using the activity

**Comment by RogerParis** [comment]
Maybe that prevents some leaking that could end up in an ANR

**Comment by RogerParis** [comment]
@angelolloqui is there a possibility that you provide us an example project that we can run and check the error?

Or would you be willing to try to reproduce the error in an example project that we provide?

Thank you very much for your help!

**Comment by angelolloqui** [comment]
Hi @RogerParis I am afraid I can not do that because the project is too big and close source, and at same time, even if I would you would struggle reproducing because it only happens in a small subset of sessions.

**Comment by RogerParis** [comment]
I see, I understand. Then I guess it makes sense that it is also hard for us to reproduce when we play around with a small simple project.

Let us know if adapting the code improves the behaviour and thanks for your feedback.

**Comment by DamienDelaire-Lydia** [comment]
Hello we are also having this issue.

**Comment by theSoenke** [comment]
Hi, after further investigation we have improved on the fix in 3.1.5 which should now solve all cases that could result in some rare circumstances in an ANR issue. The [release](https://github.com/phrase/phrase-android/releases/tag/3.2.0-rc1) also improves the performance during initialization.

**Comment by theSoenke** [comment]
Release 3.2.2 includes one more fix regarding ANRs, which fixes one case where reading the cache file could be blocked due to race condition in some rare circumstances: https://github.com/phrase/phrase-android/releases/tag/3.2.2
